api/src/main/java/org/apache/iceberg/RewriteFiles.java

```diff
@@ -84,4 +84,12 @@ RewriteFiles rewriteFiles(Set<DataFile> dataFilesToReplace, Set<DeleteFile> dele
    * @return this for method chaining
    */
   RewriteFiles validateFromSnapshot(long snapshotId);
+
+  /**
+   * Ignore the position deletes in rewrite validation. Flink upsert job only generates position deletes in the
+   * ongoing transaction, so it is not necessary to validate position deletes when rewriting.
+   *
+   * @return this for method chaining
+   */
+  RewriteFiles ignorePosDeletesInValidation();
```

Comment:
How do we know whether it is safe to use this option given that Spark can produce position delete files now and the same table can be accessed by Flink and Spark?

Have we considered exposing an option to ignore position deletes only added in the same snapshot and modifying/adding a new method to `DeleteFileIndex` to cover the new matching logic?

Reply:
Theoretically, Spark and Flink can produce deletes at the same time. While users usually use one engine to write, for example, use the Flink for upsert/CDC writes and rewrite the deletes by Spark. So I think providing an option would be better since users know how their tables are written.

Reply:
I haven't seen #4748. I'll try to check it out later today. From a quick look, I think it is similar yet a bit different from what I was proposing here. I was thinking about a new method in `DeleteFileIndex` that would only match position delete files added in later snapshots (ignoring position deletes added in the same snapshot) and then having an option to use that from the `RewriteFiles` API.

I think our assumption should be that multiple engines can write to the table (we are about to have such use cases, in fact).

Reply:
+1 to @aokolnychyi's suggestion. By ignoring delete files that were written in the same commit as a data file you're checking, you can do this safely

Reply:
Let me simulate use cases, assuming we first have the following upsert commits.
<pre>
commit-1: 
          {seq=1, [f[1], f[2], d[1]]}
commit-2: 
          {seq=1, [f[1], f[2], d[1]]},
          {seq=2, [f[3], f[4], d[2]]}
</pre>
With changes in this PR and #4748,  we could have rewrite planning and result as follows (we don't remove deletes right now):
<pre>
plan:
          task1 = {f[1], f[2], d[1]},
          task2 = {f[3], f[4], d[2]}
commit-3: 
          {seq=1, [f[5], d[1]},
          {seq=2, [f[6], d[2]]
                   
</pre>

Now, if we produce a new commit that contains new pos deletes that target f[1] and f[2].
<pre>
commit-3: 
          {seq=1, [f[1], f[2], d[1]]},
          {seq=2, [f[3], f[4], d[2]]},
          {seq=3, [d[3]]}
</pre>
With the suggested planning strategy, the result is:
<pre>    
plan: 
          task1 = {f[1], f[2], d[3]}
commit-4: 
          {seq=1, [f[1], f[2], d[1]]}, 
          {seq=2, [f[3], f[4], d[2]]}, 
          {seq=3, [f[5], d[3]]}
</pre>
It looks weird unless the assumption is that the sequence number of a position delete must be larger than its reference data files. If that is the case, I would suggest adding another option to DeleteFileIndex and RewriteFiles.

Reply:
I'm confused here, the Flink upsert job produces position deletes that target the data files produced in the same commit, so how can we ignore the deletes in the same commit when rewriting? Despite we can have later position deletes, we can not ignore the position delete that has the same sequence number.

Reply:
Well, I am no longer sure we are talking about the same use case. Let me give an example of what I meant.

```
snapshot 1 (upsert):
    f1, f2, eq1, pos1
```

Suppose `pos1` file references records in `f1`. If I understand correctly, trying to rewrite `f1` and `f2` would fail right now as there is a matching position delete. I was proposing to add a method to `RewriteFiles` that would ignore delete files for data files added in the same snapshot. This should unlock rewriting upserts and will be safe as long as `pos1` is applied whenever `f1` and `f2` are rewritten (that's how compaction works right now).

Now what should happen if a concurrent operation is committed while we are rewriting? Suppose we have another concurrent upsert.

```
snapshot 2 (upsert):
    f1, f2, eq1, pos1 (seq = 0)
    f3, f4, eq2, pos2 (seq = 1)
```

With the flag I was talking about, we will discard `pos1` but we will still validate `pos2` does not match `f1` and `f2`. Right now, we will use partition info and min/max file path to see if `pos2` matches `f1` or `f2`. It can still give us false positives but we also plan to have bloom filters, which should improve filtering.

There is also another scenario which we have to avoid.

```
snapshot 2 (spark merge-on-read):
    f1, f2, eq1, pos1 (seq = 0)
    f3, f4, pos2 (seq = 1)
```

In this case, `pos2` can indeed reference `f1` and `f2`. If that happens, we can't commit.

Reply:
@chenjunjiedada, are you interested in the first use case (without concurrent operations) or the second one (with concurrent operations)?

Reply:
> Suppose pos1 file references records in f1. If I understand correctly, trying to rewrite f1 and f2 would fail right now as there is a matching position delete. 
This rewrite should not fail IIUC, it only fails when we have another upsert snapshot that pos deletes matches `f1` or `f2`.  @jackye1995 @RussellSpitzer can you confirm this? I think this is why we think different.

Reply:
Do you mean to add bloom filter position deletes, right? Great to hear that.  Hope we can also use min/max in manifest cache and bloom filter in the data file to remove equality deletes.

Reply:
I guess you are right that the first case is handled as `validateNoNewDeletesForDataFiles` checks only new delete files added after the snapshot we used for reading.

Could you confirm the compaction you are using sets `validateFromSnapshot` correctly? If so, then the problem is that our delete matching logic may give false positives (it is currently limited to partition and file path min/max filtering). Then your solution starts to make sense but I am still concerned that it is up to the user to say if it is safe to apply this optimization and it is not something we can do when multiple engines interact with the table.

Is there any immediate optimization we can do in the delete matching logic to reduce the number of false positives? Thoughts anyone?

Reply:
> Could you confirm the compaction you are using sets validateFromSnapshot correctly? If so, then the problem is that our delete matching logic may give false positives (it is currently limited to partition and file path min/max filtering). Then your solution starts to make sense but I am still concerned that it is up to the user to say if it is safe to apply this optimization and it is not something we can do when multiple engines interact with the table.

Reply:
I agree with Anton's approach. I think we can automatically detect cases where we can do this optimization. That's much safer than adding a config.

Reply:
@aokolnychyi @rdblue Just want to double confirm, do we agree to apply bloom filtering on position delete? Applying bloom filtering on the executor side should be ok, what I'm concerned about is when planning on the driver would the bloom filtering brings heavy overhead since we have to open these small position deletes.

Reply:
Plus, we only support writer bloom filter on parquet format right now so it cannot support other format delete? Another way is to write bloom filters of deletes on puffin, not sure if that could work.

Reply:
Or we can add more info when writing position deletes, such as a new field that indicate what kind of data files it includes, the data files that have the same sequence number, or the data files that have less sequence number.  This should be the least cost AFAIK.

Reply:
@aokolnychyi @rdblue Can we write some metadata for position deletes that what sequence numbers of data files that it can reference since we can get exact data files sequence number when writing deletes. I mean add a API like `List<long> referedSeqOfDataFiles()`, a simple one might be `boolean hasSameSeqWithReferedDataFile()`